### PR TITLE
Evaluate the performance of generative models (realism and diversity)

### PR DESCRIPTION
**Pull request by JessyD**
Fix https://github.com/Project-MONAI/GenerativeModels/issues/109

### CODE REVIEW

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+# # Evaluate Realism and Diversity of the generated images
+
+# This notebook illustrates how to use the generative model package to compute the most common metrics to evaluate the performance of a generative model. The metrics that we will analyse on this tutorial are:
+#
+# - Frechet Inception Distance (FID) [1] and Maximum Mean Discrepancy (MMD) [2], two metrics commonly used to assess the realism of generated image
+#
+# - the MS-SSIM [3] and SSIM [4] used to evaluate the image diversity
+#
+# Note: We are using the RadImageNet [5] to compute the feature space necessary to compute the FID. So we need to transfom the images in the same way they were transformed when the network was trained before computing the FID.
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:22]: Typo "transform"

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+from monai.networks.layers import Act
+
+
+from monai.config import print_config
+from monai.utils import set_determinism
+
+from generative.metrics import FIDMetric, MMDMetric, MultiScaleSSIMMetric, SSIMMetric
+from generative.networks.nets import DiffusionModelUNet, PatchDiscriminator, AutoencoderKL
+from generative.networks.schedulers import DDIMScheduler
+from generative.inferers import DiffusionInferer
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:55]: It is necessary to fix the imports. It is necessary to remove the empty line

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+# +
+def subtract_mean(x: torch.Tensor) -> torch.Tensor:
+    mean = [0.406, 0.456, 0.485]
+    x[:, 0, :, :] -= mean[0]
+    x[:, 1, :, :] -= mean[1]
+    x[:, 2, :, :] -= mean[2]
+    return x
+
+def normalize_tensor(x: torch.Tensor, eps: float=1e-10) -> torch.Tensor:
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:73]: Remove unused function

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    num_channels=32,
+    in_channels=1,
+    out_channels=1,
+    kernel_size=4,
+    activation=(Act.LEAKYRELU, {"negative_slope": 0.2}),
+    norm="BATCH",
+    bias=False,
+    padding=1,
+)
+discriminator.to(device)
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:162]: Remove unused discriminator

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    unet = unet.to(device)
+else:
+    model_path = Path.cwd() / Path("tutorials/generative/2d_ldm/best_aeutoencoderkl.pth")
+    autoencoderkl.load_state_dict(torch.load(str(model_path)))
+    model_path = Path.cwd() / Path("tutorials/generative/2d_ldm/best_unet.pth")
+    unet.load_state_dict(torch.load(str(model_path)))
+
+# ## Get the real images
+
+# Simialry to the 2D LDM tutorial here we will use the MedNISTDataset, which contains images from different body parts. For easiness, here we will use only the `Hand` class. The first part of the code will get the real images from the MedNISTDataset and apply some transformation to scale the intensity of the image. Because we are evaluating the performance of the trained network, we will only use the validation split.
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:183]: Consider using
**Similar to the 2D LDM tutorial, we** will use the MedNISTDataset, which contains images from different body parts. For easiness, here we will use only the `Hand` class. The first part of the code will get the real images from the MedNISTDataset and apply some **transformations** to scale the intensity of the image. Because we are evaluating the performance of the trained network, we will only use the validation split.

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+n_synthetic_images = 3
+noise = torch.randn((n_synthetic_images, 1, 64, 64))
+noise = noise.to(device)
+scheduler.set_timesteps(num_inference_steps=50)
+unet.eval()
+
+with torch.no_grad():
+    syn_images, intermediates = inferer.sample(input_noise=noise, diffusion_model=unet,
+                                              scheduler=scheduler,save_intermediates=True,
+                                              intermediate_steps=100)
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:208]: Unused intermediates. Consider setting  `save_intermediates=False`

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+
+# ## Compute FID
+
+# The FID measures the distance between the feature vectors from the real images and those obtained from generated images. In order to compute the FID the images need to be passed into a pre-trained network to get the desired feature vectors. Although the FID is commonly computed using the Inception network, here, we used a pre-trained version of the RadImageNet to calculate the feature space. Lower FID scores indicate that the images are more similar, with a perfect score being 0 indicating that the two groups of images are identical.
+
+radnet = torch.hub.load("Warvito/radimagenet-models", model="radimagenet_resnet50", verbose=True)
+radnet.to(device)
+radnet.eval()
+
+# Here, we will load the real and generate synthetic images from noise and compute the FID of these two groups of images. Because we are generating the synthetic images on this code snippet the entire cell will take about 6 mins run and most of the this time is spent in generating the images. The loading bars show how long it will take to complete the image generation for each mini-batch.
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:225]: typo "most of this time"

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+        # Get the features for the real data
+        real_eval_feats = get_features(real_images)
+
+        # Get the features for the synthetic data
+        synth_eval_feats = get_features(syn_images)
+
+        fid = FIDMetric()
+        fid_res = fid(synth_eval_feats.to(device), real_eval_feats.to(device))
+        fid_scores.append(fid_res)
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:254]: Usually, FIDs are not computed as a mean of batches. It is evaluating the similarity between 2 distributions. consider to accumulate the features of the two groups across all batches and them compute a single FID score

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+from generative.networks.schedulers import DDIMScheduler
+from generative.inferers import DiffusionInferer
+
+print_config()
+
+
+# -
+
+# The transformations defined below are necessary in order to transform the input images in the same way that the images were
+# processed for the RadNet train.
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:63]: Typo "Radimagenet"

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+
+# Plot 3 examples from the synthetic data
+fig, ax = plt.subplots(nrows=1, ncols=3)
+for image_n in range(3):
+    ax[image_n].imshow(syn_images[image_n, 0, :, :].cpu(), cmap="gray")
+    ax[image_n].axis("off")
+
+# # Compute MMD
+
+# MMD (Maximum Mean Discrepancy) is a distance metric used to measure the similarity between two probability distributions. This metric maps the samples from each distribution to a high-dimensional feature space and calculates the distance between the mean of the features of each distribution. A smaller MMD value indicates a better match between the real and generated distributions. It is often used in combination with other evaluation metrics to assess the performance of a generative model.
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:269]: Here, I would comment that the realism of the LDM depends on the quality/realism of the AE reconstructions. For this reason you are computing the MMD between the original images and the image_recon

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+# -
+
+# # Compute MultiScaleSSIMMetric and SSIMMetric
+#
+# Both MS-SSIM and SSIM can be used as metric to evaluate the diversity.
+#
+# SSIM measures the similarity between two images based on three components: luminance, contrast, and structure. In addition, MS-SSIM is an extension of SSIM that computes the structural similarity measure at multiple scales. Both metrics can assume values between 0 and 1, where 1 indicates perfect similarity between the images.
+#
+# In this section we will compute the MS-SSIM and SSIM Meteric between the real images and those reconstructed by the AutoencoderKL.
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:296]: Make clear that this part is not related to diversity

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+                                                  intermediate_steps=100)
+
+        ms_ssim_scores.append(ms_ssim(real_images, syn_images))
+        ssim_scores.append(ssim(real_images, syn_images))
+
+ms_ssim_scores = torch.cat(ms_ssim_scores, dim=0)
+ssim_scores = torch.cat(ssim_scores, dim=0)
+
+print(f"MS-SSIM Metric: {ms_ssim_scores.mean():.7f} +- {ms_ssim_scores.std():.7f}")
+print(f"SSIM Metric: {ssim_scores.mean():.7f} +- {ssim_scores.std():.7f}")
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:352]: to measure diversity, it is necessary to compute the score between synthetic and synthetic images.

**Review by Warvito — CHANGES_REQUESTED** [review]
There are a few typos and it is necessary to fix the diversity computation. A script that might be useful is available here https://github.com/Warvito/generative_chestxray/blob/main/src/python/testing/compute_msssim_sample.py

**Comment by Warvito** [comment]
Hi @marksgraham, I saw that you are reviewing this, but I left a few comments too. I hope it is okay

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+radnet = torch.hub.load("Warvito/radimagenet-models", model="radimagenet_resnet50", verbose=True)
+radnet.to(device)
+radnet.eval()
+
+# Here, we will load the real and generate synthetic images from noise and compute the FID of these two groups of images. Because we are generating the synthetic images on this code snippet the entire cell will take about 6 mins run and most of this time is spent in generating the images. The loading bars show how long it will take to complete the image generation for each mini-batch.
+
+# +
+synth_features = []
+real_features = []
+unet.eval()
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:210]: Unet is already in evluation mode, consider remove it

File: tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+# 2. Compute the MS-SSIM and SSIM between pairs of synthetic images. This second way of computing the MS-SSIM can be used as a metric to evaluate the diversity of the synthetic images.
+#
+#
+
+# In this section we will compute the MS-SSIM and SSIM Meteric between the real images and those reconstructed by the AutoencoderKL.
+
+# +
+ms_ssim_recon_scores = []
+ssim_recon_scores = []
+autoencoderkl.eval()
```
Comment (Warvito) [inline tutorials/generative/realism_diversity_metrics/realism_diversity_metrics.py:293]: autoencoder is already in evaluation mode, cosider remove this

**Review by Warvito — CHANGES_REQUESTED** [review]
Sorry, there is just a few points that I forgot to comment. Besides that, please, execute  `./runtests.sh --autofix` to fix the format of the code.

**Review by Warvito — APPROVED** [review]
